File: app-java/src/main/java/net/mEmoZz/yts/java/ui/detail/DetailInteractor.java
package net.mEmoZz.yts.java.ui.detail;

import android.app.Activity;
import io.reactivex.Observable;
import io.reactivex.disposables.Disposable;
import net.mEmoZz.yts.java.data.network.models.MovieDetail;
import net.mEmoZz.yts.java.ui.base.BaseInteractor;
import okhttp3.ResponseBody;
import retrofit2.Response;

/**
 * Authored by Mohamed Fathy on 10 Mar, 2018.
 * Contact: muhamed.gendy@gmail.com
 */

public interface DetailInteractor extends BaseInteractor {

  interface DetailListener {

    void onDetailSubscribe(Disposable disposable);

    void onDownloadSubscribe(Disposable disposable);

    void onDetailComplete();

    void onDownloadComplete();

    void subscribeDetail(Activity context, MovieDetail model);

    void subscribeDownload(boolean isSaved);

    void onDetailError();

    void onDownloadError(Throwable throwable);

    Observable<Boolean> writeFileToDisk(Response<ResponseBody> response);
  }

  void loadMovieDetails(Activity context, long movieId, DetailListener listener);

  void downloadFile(String url, DetailListener listener);
}

File: app-kotlin/src/main/java/net/mEmoZz/yts/kotlin/ui/detail/DetailInteractor.kt
package net.mEmoZz.yts.kotlin.ui.detail

import android.app.Activity
import io.reactivex.Observable
import io.reactivex.disposables.Disposable
import net.mEmoZz.yts.kotlin.data.models.MovieDetail
import net.mEmoZz.yts.kotlin.ui.base.BaseInteractor
import okhttp3.ResponseBody
import retrofit2.Response

/**
 * Authored by Mohamed Fathy on 10 Mar, 2018.
 * Contact: muhamed.gendy@gmail.com
 */

interface DetailInteractor : BaseInteractor {

  interface DetailListener {

    fun onDetailSubscribe(disposable: Disposable)

    fun onDownloadSubscribe(disposable: Disposable)

    fun onDetailComplete()

    fun onDownloadComplete()

    fun subscribeDetail(context: Activity, model: MovieDetail)

    fun subscribeDownload(isSaved: Boolean)

    fun onDetailError()

    fun onDownloadError(throwable: Throwable)

    fun writeFileToDisk(response: Response<ResponseBody>): Observable<Boolean>
  }

  fun loadMovieDetails(context: Activity, movieId: Long, listener: DetailListener)

  fun downloadFile(url: String, listener: DetailListener)
}
